include/mesh.tcc

```diff
@@ -152,10 +169,12 @@ void Mesh<Tdim, Tvertices>::write_volumes(
     //! Write the total number of volumes (same as number of material points)
     volume_file << npoints_ << "\n";
 
-    //! Write element id and volumet
-    for (const auto& element : elements_) {
-      volume_file << id << '\t' << element->calculate_volume() << '\n';
-      ++id;
+    //! Write material point id and volume
+    for (const auto& materialpoint : materialpoints_) {
+      for (const auto& volume : materialpoint->volume()) {
```

Comment:
Why are we iterating through volumes? Shouldn't each material point have it's own volume?

Reply:
Every `material point` have `point`, and `mesh` does not have access to `point`. Thus even though each `point` has its own `volume`, we need to iterate through vector of volumes from the output in `material point`. Remember that `material point` is a collection of point.